packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol

```diff
@@ -84,71 +128,48 @@ contract FundingRateStore is FundingRateStoreInterface, Testable, Lockable {
      */
     function getFundingRateForContract(address perpetual)
         external
-        view
         override
-        nonReentrantView()
+        nonReentrant()
+        publishAndWithdrawProposal(perpetual)
         returns (FixedPoint.Signed memory)
     {
         FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(perpetual);
-
-        if (_getProposalState(fundingRateRecord.proposal) == ProposalState.Expired) {
-            return fundingRateRecord.proposal.rate;
-        } else {
-            return fundingRateRecord.rate;
-        }
+        return fundingRateRecord.rate;
     }
 
     /**
      * @notice Returns the timestamp at which the current funding rate was proposed.
-     * @dev The "current funding rate" is defined as that returned by `getFundingRateForIdentifier(id)`
      * @param perpetual Perpetual to retrieve propose time for.
      * @return propose timestamp.
      */
-    function getProposeTimeForContract(address perpetual) external view nonReentrantView() returns (uint256) {
-        return _getLatestProposeTimeForContract(perpetual);
+    function getProposeTimeForContract(address perpetual)
+        external
+        nonReentrant()
+        publishAndWithdrawProposal(perpetual)
+        returns (uint256)
+    {
+        FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(perpetual);
+        return fundingRateRecord.proposeTime;
     }
 
     /**
-     * @notice Propose a new funding rate for a perpetual. A side effect of this method is that it will
-     * overwrite the current funding rate with a pending funding rate if its liveness has expired. If this update
-     * occurs, then this method will also pay the proposer their reward for successfully updating the current rate.
+     * @notice Propose a new funding rate for a perpetual.
      * @dev This will revert if there is already a pending funding rate for the perpetual.
-     * @dev Caller must approve this this contract to spend `finalFeeBond` amount of collateral, which they can
+     * @dev Caller must approve this contract to spend `finalFeeBond` + `proposalBond` amount of collateral, which they can
      * receive back once their funding rate is published.
      * @param perpetual Perpetual contract to propose funding rate for.
      * @param rate Proposed rate.
      */
-    function propose(address perpetual, FixedPoint.Signed memory rate) external nonReentrant() {
+    function propose(address perpetual, FixedPoint.Signed memory rate)
```

Comment:
should you add a chack to see if the proposal rate != current funding rate? based off how you compute rewards the reward would be 0 but it should perhaps still be blocked. wdyt?

Reply:
Yes let's catch that